app/src/main/java/org/schabi/newpipe/player/Player.java

```diff
@@ -3364,13 +3470,229 @@ private void buildQualityMenu() {
+        if (videoFormatsCount + 1 == menu.size()) {
+            return;
+        }
+
+        menu.removeGroup(POPUP_MENU_ID_QUALITY);
+        menu.add(POPUP_MENU_ID_QUALITY, 0, Menu.NONE,
+                context.getString(R.string.auto_quality));
+
+        int menuItemId = 1;
+        for (int i = 0; i < videoTrackGroupArray.length; i++) {
+            final TrackGroup videoTrackGroup = videoTrackGroupArray.get(i);
+                final int videoTrackGroupLength = videoTrackGroup.length;
+            for (int j = 0; j < videoTrackGroupLength; j++) {
+                final Format format = videoTrackGroup.getFormat(videoTrackGroupLength - 1 - j);
+                menu.add(POPUP_MENU_ID_QUALITY, menuItemId, Menu.NONE,
+                        getResolutionStringFromFormat(context, format));
+                menuItemId++;
+            }
+        }
+
+        final TrackSelectionListener trackSelectionListener = new TrackSelectionListener(
+                context, trackSelector, videoRendererIndex, binding.qualityTextView);
+        qualityPopupMenu.setOnMenuItemClickListener(trackSelectionListener);
+        qualityPopupMenu.setOnDismissListener(this);
+        qualityPopupMenuBuiltForLivestreams = true;
+    }
+
+    /**
+     * Return if the {@link TrackGroup} of a {@link BaseTrackSelection} passed as the first
+     * parameter is one of the video {@link TrackGroup TrackGroups}.
+     * @param trackSelectionTrackGroup the {@link TrackGroup} of a {@link BaseTrackSelection} to
+     *                                 check, which cannot be null
+     * @param videoTrackGroups         the {@link TrackGroupArray} of the video
+     *                                 {@link TrackGroup TrackGroups}, which cannot be null
+     * @return {@code true} if {@code trackSelectionTrackGroup} is one of the
+     * {@code videoTrackGroups}, {@code false} otherwise
+     */
+    private boolean isSameVideoTrackGroup(
+            @NonNull final TrackGroup trackSelectionTrackGroup,
+            @NonNull final TrackGroupArray videoTrackGroups) {
+        for (int i = 0; i < videoTrackGroups.length; i++) {
+            if (videoTrackGroups.get(i) == trackSelectionTrackGroup) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Update the quality played string for livestreams and quality elements, if needed.
+     *
+     * <p>
+     * This method will also update, for the first time it's called, the default resolution,
+     * according to user settings, with
+     * {@link #setDefaultResolutionForLivestreamsIfNeeded(TrackGroupArray, int)} and will build
+     * the popup quality menu {@link #buildLiveStreamQualityMenuIfNeeded(TrackGroupArray, int)}.
+     * </p>
+     * @param trackSelections the {@link TrackSelectionArray} of ExoPlayer's track selections,
+     *                        which cannot be null
+     */
+    private void updateQualityPlayedStringAndQualityElementsForLivestreamsIfNeeded(
```

Comment:
Please use a shorter name (e.g. `updateQualityForLivestreamsIfNeeded`), the Javadoc explains the rest

Reply:
Is `updateQualityElementsForLivestreams` OK?